core/finality-grandpa/src/import.rs

```diff
@@ -458,7 +458,10 @@ impl<B, E, Block: BlockT<Hash=H256>, RA, PRA> BlockImport<Block>
 
 		match justification {
 			Some(justification) => {
-				self.import_justification(hash, number, justification, needs_justification)?;
+				if self.import_justification(hash, number, justification, needs_justification).is_err() {
```

Comment:
should we at least log the original error?

Also note that `pending_set` should be reverted in case the block with bad_justification was imported (previously we returned an `Err` here, now it's `Ok` so we fall into a different branch in finality).

Reply:
So there are three kinds of changes done to the authority set when importing a block:
- Adding pending changes which were signaled with the block;
- Enacting any forced pending changes (these are enacted on block height, not finality);
- Enacting standard pending changes if the block has a justification.

I don't think the `pending_changes` should be reverted, since the block was imported successfully so any changes signaled by it should still be added.

Reply:
I agree that we should log this tough, maybe something like:
`debug!(target: "finality", "Imported block #{} that enacts authority set change with invalid justification: {:?}, requesting justification from peers.", number, err)`

Reply:
yeah -- getting a bad justification should be treated the exact same as getting no justification in terms of mutation of internal state. The only difference is that we should report the peer who sent us a bad justification.